tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs

```diff
@@ -0,0 +1,833 @@
+
+            if (keyVaultNamespace != null || schemaMapDocument != null || kmsProviders != null || withExternalKeyVault)
+            {
+                var extraOptions = new Dictionary<string, object>()
+                {
+                    { "mongocryptdSpawnPath", Environment.GetEnvironmentVariable("MONGODB_BINARIES") ?? string.Empty }
+                };
+
+                var schemaMap = GetSchemaMapIfNotNull(schemaMapDocument);
+
+                if (kmsProviders == null)
+                {
+                    kmsProviders = new ReadOnlyDictionary<string, IReadOnlyDictionary<string, object>>(new Dictionary<string, IReadOnlyDictionary<string, object>>());
+                }
+
+                var autoEncryptionOptions = new AutoEncryptionOptions(
+                    keyVaultNamespace: keyVaultNamespace,
+                    kmsProviders: kmsProviders,
+                    schemaMap: schemaMap,
+                    extraOptions: extraOptions);
+
+                if (withExternalKeyVault)
+                {
+                    var externalKeyVaultClientSettings = DriverTestConfiguration.GetClientSettings().Clone();
+                    externalKeyVaultClientSettings.Credential = MongoCredential.FromComponents(null, null, "fake-user", "fake-pwd");
+                    var externalKeyVaultClient = new MongoClient(externalKeyVaultClientSettings);
+                    autoEncryptionOptions = autoEncryptionOptions.With(keyVaultClient: externalKeyVaultClient);
+                }
+                mongoClientSettings.AutoEncryptionOptions = autoEncryptionOptions;
+            }
+
+            return new DisposableMongoClient(new MongoClient(mongoClientSettings));
+        }
+
+        private Dictionary<string, BsonDocument> GetSchemaMapIfNotNull(BsonDocument schemaMapDocument)
+        {
+            Dictionary<string, BsonDocument> schemaMap = null;
+            if (schemaMapDocument != null)
+            {
+                var element = schemaMapDocument.Single();
+                schemaMap = new Dictionary<string, BsonDocument>
+                    {
+                        { element.Name, element.Value.AsBsonDocument }
+                    };
+            }
+            return schemaMap;
+        }
+
+        private void Insert(
+            IMongoCollection<BsonDocument> collection,
+            bool async,
+            params BsonDocument[] documents)
+        {
+            if (async)
+            {
+                collection
+                    .InsertManyAsync(documents)
+                    .GetAwaiter()
+                    .GetResult();
+            }
+            else
+            {
+                collection.InsertMany(documents);
+            }
+        }
+
+        public class JsonFileReader : EmbeddedResourceJsonFileReader
+        {
+            #region static
+            private static JsonFileReader __instance;
+            public static JsonFileReader Instance => __instance ?? (__instance = new JsonFileReader());
```

Comment:
Maybe use `Lazy<T>` for thread safety?

Doesn't currently matter because we disable parallelism in XUnit... but would be more correct.

Reply:
done